File: src/utils/tag-utils.ts
import { Blueprint } from '@/types';

/**
 * Convert a tag string or array to a tag array.
 * @param tag Either a pipe-delimited string (e.g. "foo|bar") or a list of strings
 * @returns List of tag components
 */
export function tagToArray(tag: string | string[]): string[] {
    if (typeof tag === 'string') {
        return tag.split('|');
    }
    return tag;
}

/**
 * Convert a tag array to a pipe-delimited string.
 * @param tagArray List of tag components
 * @returns Pipe-delimited string
 */
export function arrayToTag(tagArray: string[]): string {
    return tagArray.join('|');
}

/**
 * Convert a tag dictionary to use pipe-delimited strings.
 * @param tagDict Dictionary containing a 'tag' key with array value
 * @returns Dictionary with 'tag' value converted to pipe-delimited string
 */
export function convertTagDict(tagDict: { tag: string[] }): { tag: string } {
    return {
        ...tagDict,
        tag: arrayToTag(tagDict.tag)
    };
}

/**
 * Get other blueprint tags from config selections, parent blueprint, and peer parts
 * @param configSelections - Current configuration selections
 * @param currentPartName - Name of the current part
 * @param parentBlueprint - The parent blueprint containing this config
 * @returns Object containing parent tags and sibling selections
 */
export function getOtherBlueprintTags(
  configSelections: Record<string, Blueprint>,
  currentPartName: string,
  parentBlueprint?: Blueprint
): { parentTags: string[]; siblingSelections: { partName: string; tags: string[] }[] } {
  const parentTags: string[] = [];
  const siblingSelections: { partName: string; tags: string[] }[] = [];
  
  // Add tags from parent blueprint
  if (parentBlueprint) {
    parentTags.push(...parentBlueprint.tags);
  }
  
  // Add tags from other selected parts as sibling selections
  Object.entries(configSelections).forEach(([partName, bp]) => {
    if (partName !== currentPartName) {
      siblingSelections.push({
        partName,
        tags: bp.tags
      });
    }
  });
  
  return { parentTags, siblingSelections };
}

/**
 * Swap tags by filtering out a specific tag type
 * @param blueprint - The blueprint containing tags to filter
 * @param tagType - The tag type to filter out (e.g., 'texture', 'size', 'connection')
 * @returns Array of tags that don't start with the specified tag type
 */
export function swapTagsByType(blueprint: Blueprint, tagType: string): string[] {
  return blueprint.tags.filter(t => !t.startsWith(tagType));
} 

File: openforge/db/sql/tag_utils.py
from typing import Union, List

def tag_to_array(tag: Union[str, List[str]]) -> List[str]:
    """Convert a tag string or array to a tag array.
    
    Args:
        tag: Either a pipe-delimited string (e.g. "foo|bar") or a list of strings
        
    Returns:
        List of tag components
    """
    if isinstance(tag, list):
        return tag
    if isinstance(tag, str):
        return tag.split("|")
    return list(tag)

def array_to_tag(tag_array):
    return "|".join(str(x) for x in tag_array)

def convert_tag_dict(tag_dict):
    if tag_dict is None:
        return None
    if 'tag' in tag_dict and isinstance(tag_dict['tag'], list):
        tag_dict['tag'] = array_to_tag(tag_dict['tag'])
    return tag_dict

def process_tag(tag, process_func):
    """Process a tag with consistent error handling for different formats.
    
    Args:
        tag: Either a list of strings or a pipe-delimited string
        process_func: Function to apply to the processed tag array
        
    Returns:
        Result of applying process_func to the tag array
        
    Raises:
        TypeError: If tag is neither list nor string
    """
    if isinstance(tag, list):
        # Old format: ["shape", "floor"] -> process directly
        return process_func(tag)
    elif isinstance(tag, str):
        # New format: "shape|floor" -> convert to array, then process
        tag_array = tag_to_array(tag)
        return process_func(tag_array)
    else:
        # Fail fast for unexpected tag types
        raise TypeError(f"Unsupported tag type {type(tag).__name__}: {tag}")
